client/state/posts/actions.js

```diff
@@ -252,7 +258,37 @@ export function savePost( siteId, postId = null, post ) {
  * @return {Function}        Action thunk
  */
 export function trashPost( siteId, postId ) {
-	return savePost( siteId, postId, { status: 'trash' } );
+	return dispatch => {
+		const post = { status: 'trash' };
+
+		dispatch( {
+			type: POST_SAVE,
+			siteId,
+			postId,
+			post,
+		} );
```

Comment:
To confirm my understanding here... this is the "optimistic update" that causes the post to immediately disappear from the list, and then some of the "shadow item" code works around that.  But we're not ready to remove the optimistic update because we don't have the in progress and undo UX for the posts list yet.

Reply:
Yes. First, I didn't want to make any changes that would affect `PostTypeList` as well. I'm afraid I would have to do a long chain of cascading code updates and never got this PR finished -- it already inflated a lot from the original intent 🙂

Second, I don't want to touch anything related to `PostQueryManager` right now -- that's a component I still don't fully understand and I remember it caused a lot of trouble during the `PostTypeList` project when trying to optimize the Redux dispatches there.

In this PR, I'm looking for a reasonable tradeoff between doing some horrible hack and "doing it properly", which would require rewriting large parts of `PostQueryManager`, `PostTypeList`, the page list UI etc. Let's do that in smaller steps.

Even if the "optimistic update" was removed and Redux state was updated only on `POST_SAVE_SUCCESS`, the "shadow item" logic wouldn't change very much. I still need to continue displaying the item after it's removed from the list, because I need to show the success notice for a few seconds, and the "Undo" action, too.